experiments/experiment_prototype.py

```diff
@@ -1,385 +1,1111 @@
         """
 
-        if (self.cponum<1):
-            errmsg="Error: No objects in control program"
-            sys.exit(errmsg) # REVIEW 6 Add a todo for error handling. Perhaps use exceptions instead.
+        if self.num_slices < 1:
+            errmsg = "Error: Invalid num_slices less than 1"
+            raise ExperimentException(errmsg)
 
-        #TODO: somehow check if self.cpid is not unique
+        # TODO: somehow check if self.cpid is not unique - incorporate known cpids from git repo?
 
-        for cpo in range(self.cponum):
-            selferrs=selfcheck_piece(self.cpo_list[cpo],self.config)
-            if (not selferrs): # REVIEW #39 unneeded parenthesis
+        for a_slice in self.slice_ids:
+            selferrs = self.check_slice(self.slice_dict[a_slice])
+            if not selferrs:
                 # If returned error dictionary is empty
                 continue
-            errmsg="Self Check Errors Occurred with Object Number : {} \nSelf \
-                Check Errors are : {}".format(cpo, selferrs)
-            sys.exit(errmsg) # REVIEW 6 Add a todo for error handling. Perhaps use exceptions instead.
-        else: # no break (exit)
-            print "No Self Check Errors. Continuing..."
+            errmsg = "Self Check Errors Occurred with slice Number : {} \nSelf \
+                Check Errors are : {}".format(a_slice, selferrs)
+            raise ExperimentException(errmsg)
+
+        print("No Self Check Errors. Continuing...")
+
         return None
 
+    def check_slice(self, exp_slice):
+        """
+        This is the first test of the dictionary in the experiment done to ensure values in this 
+        slice make sense. This is a self-check to ensure the parameters (for example, txfreq, 
+        antennas) are appropriate. All fields should be full at this time (whether filled by the 
+        user or given default values in set_slice_defaults).
+        :param: exp_slice: a slice to check
+        """
+        error_count = 0
+        error_dict = {}
+
+        options = self.options
+
+        for param in self.slice_keys:
+            try:
+                 assert param in exp_slice.keys()
+            except AssertionError:
+                if param == 'txfreq' and exp_slice['clrfrqflag']:
+                    pass
+                elif param == 'rxfreq' and not exp_slice['rxonly']:
+                    pass
+                elif param == 'clrfrqrange' and not exp_slice['clrfrqflag']:
+                    pass
+                else:
+                    errmsg = "Slice {} is missing Necessary Parameter {}".format(
+                        exp_slice['slice_id'], param)
+                    raise ExperimentException(errmsg)
+                    # set defaults if possible
+            try:
+                assert param is not None
+            except AssertionError:
+                pass  # TODO may want to check certain params are not None
+
+        if error_dict:  # if not empty
```

Comment:
we think this if statement will do nothing at this point since error_dict has not been added to yet